extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -128,235 +123,268 @@ export class Bind {
   }
 
   /**
-   * Scans children for attributes that conform to bind syntax and returns
-   * all bindings.
+   * Scans `body` for attributes that conform to bind syntax and returns
+   * a tuple containing bound elements and binding data for the evaluator.
    * @param {!Element} body
-   * @return {{bindings: !Array<BindingDef>, expressions: !Array<string>}}
+   * @return {{
+   *   boundElements: !Array<BoundElementDef>,
+   *   evaluatees: !Array<./bind-evaluator.EvaluateeDef>,
+   * }}
    * @private
    */
   scanForBindings_(body) {
     // TODO(choumx): Chunk if taking too long in a single frame.
 
-    const bindings = [];
-    const expressions = [];
+    /** @type {!Array<BoundElementDef>} */
+    const boundElements = [];
+    /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
+    const evaluatees = [];
+
     const elements = body.getElementsByTagName('*');
     for (let i = 0; i < elements.length; i++) {
-      const el = elements[i];
-      const attributes = el.attributes;
+      const element = elements[i];
+      const attributes = element.attributes;
+
+      const bindings = [];
       for (let j = 0; j < attributes.length; j++) {
-        const binding = this.bindingForAttribute_(attributes[j], el);
+        const binding = this.bindingForAttribute_(attributes[j], element);
         if (binding) {
           bindings.push(binding);
-          expressions.push(binding.expression);
+          evaluatees.push({
+            tagName: element.tagName,
+            property: binding.property,
+            expressionString: binding.expression,
+          });
         }
       }
+      if (bindings.length > 0) {
+        boundElements.push({element, bindings});
+      }
     }
-    return {bindings, expressions};
+    return {boundElements, evaluatees};
   }
 
   /**
    * Returns a struct representing the binding corresponding to the
    * attribute param, if applicable.
    * @param {!Attr} attribute
-   * @param {!Element} element
+   * @param {!Element} unusedElement
    * @return {?BindingDef}
    * @private
    */
-  bindingForAttribute_(attribute, element) {
-    // TODO(choumx): Only allow binding to attributes allowed by validator.
-
+  bindingForAttribute_(attribute, unusedElement) {
     const name = attribute.name;
-    if (name.length > 2) {
-      if (name[0] === '[' && name[name.length - 1] === ']') {
-        return {
-          property: name.substr(1, name.length - 2),
-          expression: attribute.value,
-          element,
-        };
-      }
+    if (name.length > 2 && name[0] === '[' && name[name.length - 1] === ']') {
```

Comment:
Did we confirm if these attribute names valid HTML? I've been trying to set it via `Element.setAttribute` with no luck.